lib/resty/worker/events.lua

```diff
@@ -406,30 +407,27 @@ _M.poll = function()
   return _M.poll()
 end
 
--- executes a polling loop, and reschedules the polling timer
+-- executes a polling loop
 local do_timer
 do_timer = function(premature)
-  if premature then
-    _M.post(_M.events._source, _M.events.stopping)
-  end
-
-  local ok, err = _M.poll()
-  if not ok then
-    log(ERR, "worker-events: timer-poll returned: ", err)
-  end
+  while true do
+    if premature then
+      _M.post(_M.events._source, _M.events.stopping)
+    end
 
-  if _interval ~= 0 and not premature then
-    ok, err = new_timer(_interval, do_timer)
+    local ok, err = _M.poll()
     if not ok then
-      if err == "process exiting" then
-        _M.post(_M.events._source, _M.events.stopping)
-      end
-      err = "failed to create timer: " .. tostring(err)
-      log(ERR, "worker-events: ", err)
-      return nil, err
+      log(ERR, "worker-events: timer-poll returned: ", err)
+    end
+
+    if _interval ~= 0 and not premature then
+      sleep(_interval)
```

Comment:
the problem with this sleep is that it will not get interrupted when the system exits. So if the timer is set as a fallback; say 30 seconds. And the system exits, then probably before the 30 seconds are hit, the system will do a forced shutdown. And in that case the stop event (on line 415 above) will never be send.

Reply:
Considering that this lib is a singelton, and only uses a single timer, it should not be a burden on the timer resources of an openresty based system.

So considering the negative impact on functionality, I think we should not be merging this.

Reply:
See this:
```
worker_processes  1;

events {
    worker_connections  1024;
}

http {
    init_worker_by_lua_block {
        ngx.timer.at(0, function(premature)
            while not premture do
                ngx.sleep(100)
                premature = ngx.worker.exiting()
            end
        end)
    }

    server {
        listen       8080;
        server_name  localhost;
        default_type text/html; 
        location / {
            content_by_lua '
                ngx.say("<html><body>Hello!</body></html>")
            ';
        }
    }
}
```
Start it with:
```
$ openresty -c timer.conf -p timer
```
And then run stop. The sleep will get interrupted.

Reply:
Though to be noted, the sleeping thread is not resumed, it is just killed. So with this you are not able to run the cleanup.

Reply:
> Though to be noted, the sleeping thread is not resumed, it is just killed. So with this you are not able to run the cleanup.

I need to clarify this. Signal `stop` just kills these threads without letting them know anything. Signal `reload` and `quit` will call `timer` with premature, and also `resume the loop` with `ngx.worker.exiting() == true`. So there should be no introduced issues with this PR.

Reply:
Here is the full config that I used with different signals:
```
worker_processes  1;

events {
    worker_connections  1024;
}

http {
    init_worker_by_lua_block {
        local timer
        timer = function(premature)
            ngx.log(ngx.ERR, "CLEANUP TIMER STARTED")
            if premature then
                ngx.log(ngx.ERR, "PREMATURELY CLOSING TIMER")
                return
            end
            ngx.log(ngx.ERR, "CLEANUP TIMER ENDED")
            local ok, err = ngx.timer.at(5, timer)
            if not ok then
                ngx.log(ngx.ERR, "CLEANUP TIMER ERROR: ", err)
            end
        end

        ngx.timer.at(0, function(premature)
            local ok, err = ngx.timer.at(5, timer)

            if not ok then
                ngx.log(ngx.ERR, "ERROR: ", err)
            end

            ngx.log(ngx.ERR, "LOOP TIMER STARTED")

            while not premature do
                ngx.log(ngx.ERR, "sleeping started at: ", ngx.now())
                ngx.sleep(5)
                ngx.log(ngx.ERR, "sleeping iterrupted at: ", ngx.now())
                premature = ngx.worker.exiting()
            end

            ngx.log(ngx.ERR, "PREMATURELY CLOSING LOOP TIMER")
        end)
    }

    server {
        listen       8080;
        server_name  localhost;
        default_type text/html;
        location / {
            content_by_lua '
                ngx.say("<html><body>Hello!</body></html>")
            ';
        }
    }
}
```

Reply:
Damnit, I stand corrected once more. If we sleep for 100 seconds, that makes Nginx graceful exit with `quit` wait until that sleep ends. I changed the PR so that now it sleeps in 1 second blocks.